### PR TITLE
Problem: outdated c-binding example (wip #1845)

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: update sync json-rpc protocol  and added automatic `chain.h` modification for gcc

### CODE REVIEW

File: cro-clib/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,9 @@ edition = "2018"
 
 [lib]
 name = "cro_clib"
-crate-type = ["staticlib"]
+#in ios, android, use staticlib
+#crate-type = ["staticlib"]
```
Comment (yihuang) [inline cro-clib/Cargo.toml:10]: remove this line, might confuse the sed script in `docker/build.sh`

Comment (leejw51crypto) [reply cro-clib/Cargo.toml:10]: ok, i'll check build.sh

Comment (leejw51crypto) [reply cro-clib/Cargo.toml:10]: also, will inspect travis

**Comment by leejw51crypto** [comment]
rebased

File: cro-clib/Cargo.toml [hunk 2]
```diff
@@ -6,7 +6,8 @@ edition = "2018"
 
 [lib]
 name = "cro_clib"
-crate-type = ["staticlib"]
+#in ios, android, use staticlib instead of cdylib
+crate-type = ["cdylib"]
```
Comment (tomtau) [inline cro-clib/Cargo.toml:10]: isn't staticlib better default?

File: cro-clib/examples/Makefile [hunk 3]
```diff
@@ -1,17 +1,22 @@
 all: build run
 build:
+# for rocksdb link errors
+	export AR=/usr/bin/ar
 	cargo build
+	sed   's/Option_ProgressCallback/ProgressCallback/' -i  ../chain.h
```
Comment (tomtau) [inline cro-clib/examples/Makefile:6]: could this be patched in https://github.com/eqrion/cbindgen/blob/master/docs.md#cbindgentoml
there's "rename" or include/exclude for export?

File: cro-clib/Cargo.toml (hunk 2 above)
Comment (leejw51crypto) [reply cro-clib/Cargo.toml:10]: for staticlib, link is OK, but when i run example on ubuntu, its crashes (seg-fault)

File: cro-clib/examples/Makefile (hunk 3 above)
Comment (leejw51crypto) [reply cro-clib/examples/Makefile:6]: i'll check

File: cro-clib/Cargo.toml (hunk 2 above)
Comment (tomtau) [reply cro-clib/Cargo.toml:10]: where does it segfault? can you do strace of what's the last thing it calls?

Comment (tomtau) [reply cro-clib/Cargo.toml:10]: or run it in a debugger?

Comment (leejw51crypto) [reply cro-clib/Cargo.toml:10]: i tried again, it's working fine with ["staticlib"], i'll change back

**Comment by leejw51crypto** [comment]
removed `sed` post-processing,
changed to struct function wrapper
now, testing

**Comment by leejw51crypto** [comment]
ok, i'll check clippy issues

**Pull request merged**